packages/token/src/tests/erc20/test_erc20_permit.cairo

```diff
@@ -0,0 +1,483 @@
+fn test_invalid_sig_bad_metadata_name() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
+
+    let mut modified_data = data;
+    modified_data.metadata_name = 'ANOTHER_NAME';
+    let nonce = state.nonces(owner);
+    let signature = prepare_permit_signature(modified_data, nonce);
+    state.permit(owner, spender, amount, deadline, signature);
+}
+
+#[test]
+#[should_panic(expected: 'ERC20: invalid permit signature')]
+fn test_invalid_sig_bad_metadata_version() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
+
+    let mut modified_data = data;
+    modified_data.metadata_version = 'ANOTHER_VERSION';
+    let nonce = state.nonces(owner);
+    let signature = prepare_permit_signature(modified_data, nonce);
+    state.permit(owner, spender, amount, deadline, signature);
+}
+
+#[test]
+#[should_panic(expected: 'ERC20: invalid permit signature')]
+fn test_invalid_sig_bad_signing_key() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
+
+    let mut modified_data = data;
+    modified_data.key_pair = constants::stark::KEY_PAIR_2();
+    let nonce = state.nonces(owner);
+    let signature = prepare_permit_signature(modified_data, nonce);
+    state.permit(owner, spender, amount, deadline, signature);
+}
+
+#[test]
+#[should_panic(expected: 'ERC20: invalid permit signature')]
+fn test_invalid_sig_bad_chain_id() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
+
+    let mut modified_data = data;
+    modified_data.chain_id = 'ANOTHER_CHAIN_ID';
+    let nonce = state.nonces(owner);
+    let signature = prepare_permit_signature(modified_data, nonce);
+    state.permit(owner, spender, amount, deadline, signature);
+}
+
+#[test]
+#[should_panic(expected: 'ERC20: invalid permit signature')]
+fn test_invalid_sig_bad_revision() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
+
+    let mut modified_data = data;
+    modified_data.revision = 'ANOTHER_REVISION';
+    let nonce = state.nonces(owner);
+    let signature = prepare_permit_signature(modified_data, nonce);
+    state.permit(owner, spender, amount, deadline, signature);
+}
+
+//
+// Expired signature
+//
+
+#[test]
+#[should_panic(expected: 'ERC20: expired permit signature')]
+fn test_invalid_sig_bad_deadline() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
```

Comment:
This is an example of a test that doesn't use the `DualCaseAccountMock`, should we decouple the `setup_account` logic from the `setup` one?

Reply:
This particular test case doesn't require the account mock to be deployed. At the same time, I don't think we should decouple it, since this way we'll lose consistency over the test cases and we'll have to introduce an additional setup function that will be used in this only place